### ISSUE TITLE
maximum/minimum over dimension fails

### ISSUE THREAD

**Issue by nalimilan** [JuliaData/CategoricalArrays.jl#178]
```julia
julia> maximum(categorical([1 2; 3 4]), dims=1)
ERROR: ArgumentError: CategoricalValue objects with different pools cannot be tested for order
```

We would need an API so that `maximum`/`minimum` can allocate a new array which shares the pool with the original array. Though even that wouldn't be ideal, as modifying the pool after the fact would have surprising results. So the pool should be copied at the end of the process. Maybe the only solution is to override `reducedim_init`.

**Comment by bkamins** [comment]
We have the problem with modifying the pool already now:
```
julia> x = categorical([1,2,3,4])
4-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3
 4

julia> m = maximum(x)
CategoricalValue{Int64,UInt32} 4

julia> x[4] = 1
1

julia> levels!(x, [3,2,1])
4-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3
 1

julia> m
Error showing value of type CategoricalValue{Int64,UInt32}:
ERROR: BoundsError: attempt to access 3-element Array{Int64,1} at index [4]
```

So I think that reusing the pool in dimensional reductions should be acceptable in the same way. But the issue is far from easy to decide (we are getting into another complex issue similarly to combining levels).

**Comment by nalimilan** [comment]
Yeah, but a categorical value is less likely to be kept around for a long time (typically you'll store it inside another array or convert it). OTC a categorical array resulting from a reduction over dimensions can be stored for a longer period I guess. Also it's more obvious/acceptable that a categorical value refers to its parent array.

**Comment by bkamins** [comment]
I guess we have to do a custom reducer then.

Additionally I would consider to add a branch in `isless` that in case when `pool(x) !== pool(y)` checks if actually the pools are equal. It will be slow, but it is better than simply throwing an error.

Then maybe we could additionally have `pool(x) !== pool(y)`  in a checking bounds branch so `@inbounds isless(x,y)` would not check the `pool(x) !== pool(y)` condition?

**Comment by bkamins** [comment]
So what is currently an idiomatic way to say:

> vectors `x` and `y` have the same pool and their elements can be compared?

Because if I understand your current design correctly (admittedly I do not know all the details yet) it is not possible.

While:

> elements of categorical vector `x` and non-categorical vector `y` are comparable based on the order of `x`.

right?

**Comment by nalimilan** [comment]
No, it's not possible now unfortunately... And yes, it's kind of inconsistent to allow comparisons with non categorical values, which is as slow as with categorical values from a different pool. So maybe we should allow both even if that's slow and hope the compiler can make this fast at some point.
